### PR TITLE
feat: Promote FormField from Preview to Main

### PR DESCRIPTION
**Pull request by mannycarrera4**
## Summary

Fixes: https://github.com/Workday/canvas-kit/issues/1934

## Release Category
Components

### Release Note
We've promoted FormField from [Preview](#preview) to [Main](#main). The following changes have been
made to provide more flexibility and better explicit components when using inputs.

- The orientation prop on the `FormField` component will be updated to accept the following values:
  `vertical`, `horizontalStart`, and `horizontalEnd`. `horizontal` will still be accepted as a value
  in v12, but it will be deprecated and slated for removal in a future major release. These changes
  are intended to provide more flexibility with label alignments on `FormField` elements.

> **Note**: The horizontal alignment of start and end are relative to its container, meaning start
> and end match the flex property of `flex-start` and `flex-end`. This is especially applicable for
> moving between RTL (right-to-left) and LTR (left-to-right) languages.

> **Note:** Orientation "horizontal" has been deprecated. You will see a console warn message
> suggesting to update your types and usage to `horizontalStart`, `horizontalEnd` or `vertical`.

- `useFormFieldModel`: `orientation` has been added back into `useFormFieldModel`. While we still
  support compat mode due to
  [style merging issues](https://github.com/Workday/canvas-kit/discussions/2893), having orientation
  on the model allows for proper styling of sub components.

- Styles clean up. `FormField.Hint` and `FormField.Label` where using `margin` for spacing. We've
  updated styles so that the containing element uses `gap` for proper spacing.

### BREAKING CHANGES
The newly promoted `FormField` is a
[compound component](https://workday.github.io/canvas-kit/?path=/docs/guides-compound-components--page).
Due to the different APIs of the component, this change is **not codemodable**. The following shows
an example of how to **update** your code to match the new compound component API.

```tsx
// v11 FormField in Main
<FormField
  error={FormField.ErrorType.Alert}
  labelPosition={FormField.LabelPosition.Left}
  useFieldSet={true}
  required={true}
  hintId="hint-alert"
  hintText="Please enter a valid email."
  label="Email"
>
  <TextInput onChange={handleChange} value={value} />
</FormField>


// v12 Newly promoted FormField from Preview to Main
<FormField
  error="error"
  orientation="horizontalStart"
  isRequired={true}
>
	<FormField.Label>Email</FormField.Label>
	<FormField.Field>
		<FormField.Input as={TextInput} onChange={handleChange} value={value} />
		<FormField.Hint>Please enter a valid email.</FormField.Hint>
	</FormField.Field>
</FormField>
```

**Noticeable changes:**

- `error` prop takes in the following values: `"error" | "alert"`.
- `labelPosition` becomes `orientation` with the following values:
  `"horizontal" | "horizontalStart" | "horizontalEnd" | "vertical"`.
- `useFieldSet` is no longer valid. If you want to group inputs, use
  [`FormFieldGroup`](#form-field-group) component.
- `required` becomes `isRequired`.
- `hintId` is no longer needed. The component handles associating the hint text, label and input
  automatically. If you wish to have a unique `id`, you can add one onto the `FormField` element.
- `hintText` is no longer a valid prop. Use `FormField.Hint` sub component.
- `errorLabel` and `alertLabel` are no longer valid props. Customizing your hint text inside of
  `FormField.Hint`.
- `label` is no longer a valid prop. Use `FormField.Label` sub component to render your label text.
- Instead of rendering an input, ensure you use `FormField.Input` with the `as` prop. This ensures
  proper error handling and aria attributes for accessibility.

---

## Checklist

- [ ] MDX documentation adheres to Canvas Kit's [Documentation Guidelines](https://workday.github.io/canvas-kit/?path=/docs/guides-documentation-guidelines--page)
- [ ] Label `ready for review` has been added to PR

## For the Reviewer

- [ ] PR title is short and descriptive
- [ ] PR summary describes the change (Fixes/Resolves linked correctly)
- [ ] PR Release Notes describes additional information useful to call out in a release message or removed if not applicable
- [ ] Breaking Changes provides useful information to upgrade to this code or removed if not applicable

## Areas for Feedback? (optional)

- [ ] Code
- [ ] Documentation
- [ ] Testing
- [ ] Codemods

### CODE REVIEW

File: cypress/component/SelectPreview.spec.tsx [hunk 1]
```diff
@@ -866,7 +868,7 @@ describe('Select', () => {
       );
     });
 
-    context(`given the "Portal Test" story is rendered`, () => {
+    context.skip(`given the "Portal Test" story is rendered`, () => {
```
Comment (mannycarrera4) [inline cypress/component/SelectPreview.spec.tsx:871]: Skipping this test :(

File: modules/docs/mdx/11.0-UPGRADE-GUIDE.mdx [hunk 2]
```diff
@@ -609,7 +609,7 @@ are applied to `FormField` in [Preview](#preview) in order to match the API from
 
 ```tsx
 // v10 FormField in Preview
-import {FormField} from '@workday/canvas-kit-preview-react/form-field';
+import {FormField} from '@workday/canvas-kit-react/form-field';
```
Comment (mannycarrera4) [inline modules/docs/mdx/11.0-UPGRADE-GUIDE.mdx:612]:
```suggestion
import {FormField} from '@workday/canvas-kit-preview-react/form-field';
```

File: modules/docs/mdx/11.0-UPGRADE-GUIDE.mdx [hunk 3]
```diff
@@ -619,7 +619,7 @@ import {TextInput} from '@workday/canvas-kit-react/text-input';
 </FormField>;
 
 // v11
-import {FormField} from '@workday/canvas-kit-preview-react/form-field';
+import {FormField} from '@workday/canvas-kit-react/form-field';
```
Comment (mannycarrera4) [inline modules/docs/mdx/11.0-UPGRADE-GUIDE.mdx:622]:
```suggestion
import {FormField} from '@workday/canvas-kit-preview-react/form-field';
```

File: modules/preview-react/_examples/stories/mdx/examples/Density.tsx [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import {FormField, FormFieldGroup} from '@workday/canvas-kit-preview-react/form-field';
+import {FormField, FormFieldGroup} from '@workday/canvas-kit-react/form-field';
```
Comment (mannycarrera4) [inline modules/preview-react/_examples/stories/mdx/examples/Density.tsx:2]: I should move these stories out of preview after the promotion. Also reference this example in the upgrade guide

File: modules/codemod/lib/v12/spec/renameFormFieldHorizontal.spec.ts [hunk 5]
```diff
@@ -60,13 +60,13 @@ describe('rename horizontal', () => {
...
     const input = stripIndent`
-        import {FormField} from '@workday/canvas-kit-preview-react/form-field'
+        import {FormField} from '@workday/canvas-kit-react/form-field'
         const StyledForm = styled(FormField)({color: "#000"});
         <StyledForm orientation="horizontal" />
     `;
 
     const expected = stripIndent`
-        import {FormField} from '@workday/canvas-kit-preview-react/form-field'
+        import {FormField} from '@workday/canvas-kit-react/form-field'
```
Comment (mannycarrera4) [inline modules/codemod/lib/v12/spec/renameFormFieldHorizontal.spec.ts:69]: fix tests

File: modules/labs-react/combobox/lib/Combobox.tsx [hunk 6]
```diff
@@ -427,7 +427,6 @@ export const Combobox = ({
         React.InputHTMLAttributes<HTMLInputElement> & {ref: React.Ref<HTMLInputElement>}
     > & {css: CSSObject} = {
       type: 'text',
-      id: componentId,
```
Comment (mannycarrera4) [inline modules/labs-react/combobox/lib/Combobox.tsx:430]: this is being set by the formfield component.

File: modules/codemod/lib/v12/index.ts [hunk 7]
```diff
@@ -1,11 +1,13 @@
 import {Transform} from 'jscodeshift';
 
 import renameFormFieldHorizontal from './renameFormFieldHorizontal';
+import promoteFormField from './promoteFormField';
 
 const transform: Transform = (file, api, options) => {
   // These will run in order. If your transform depends on others, place yours after dependent transforms
   const fixes = [
     // add codemods here
+    promoteFormField,
```
Comment (mannycarrera4) [inline modules/codemod/lib/v12/index.ts:10]: Promote formfield first, then make the change to horizontal

File: modules/react/form-field/README.md [hunk 8]
```diff
@@ -1,5 +1,8 @@
-# Canvas Kit Form Field
+# Canvas Kit React Form Field
 
 View the
-[documentation for Form Field](https://workday.github.io/canvas-kit/?path=/docs/components-inputs-form-field-react--basic)
+[documentation for Form Field](https://workday.github.io/canvas-kit/?path=/docs/preview-inputs-form-field-react--custom)
```
Comment (mannycarrera4) [inline modules/react/form-field/README.md:4]:
```suggestion
[documentation for Form Field](https://workday.github.io/canvas-kit/?path=/docs/components-inputs-form-field--docs)
```

File: cypress/component/SelectPreview.spec.tsx (hunk 1 above)
Comment (josh-bagwell) [reply cypress/component/SelectPreview.spec.tsx:871]: Was this an issue? Was it not passing?

Comment (mannycarrera4) [reply cypress/component/SelectPreview.spec.tsx:871]: yeah it was failing

File: modules/preview-react/_examples/stories/mdx/examples/Density.tsx (hunk 4 above)
Comment (mannycarrera4) [reply modules/preview-react/_examples/stories/mdx/examples/Density.tsx:2]: done

File: cypress/component/SelectPreview.spec.tsx (hunk 1 above)
Comment (josh-bagwell) [reply cypress/component/SelectPreview.spec.tsx:871]: Cool. Yeah we're going to have to redo a lot of the specs as component runner seems to be causing quite a bit of issues

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx [hunk 9]
```diff
@@ -58,9 +59,8 @@ A note to the reader:
 
 In v12, all the components listed in this guide have started using our new
 [Canvas Tokens Web](https://workday.github.io/canvas-tokens/?path=/docs/docs-getting-started--docs).
-
-In v12 you must add `@workday/canvas-tokens-web` to ensure our components are properly styled and
-the variables are defined. For more information on installing and using, please reference our tokens
+You must add `@workday/canvas-tokens-web` to ensure our components are properly styled and the
```
Comment (josh-bagwell) [inline modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:62]: Did we want to bold this like above? 
```suggestion
**You must** add `@workday/canvas-tokens-web` to ensure our components are properly styled and the
```

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx [hunk 10]
```diff
@@ -226,50 +228,121 @@ The following changes have been made to the API:
...
+
+- `error` prop takes in the following values: `"error" | "alert"`.
+- `labelPosition` becomes `orientation` with the following values:
+  `"horizontal" | "horizontalStart" | "horizontalEnd" | "vertical"`.
+- `useFieldSet` is no longer valid. If you want to group inputs, use
+  [`FormFieldGroup`](#form-field-group) component.
+- `required` becomes `isRequired`.
+- `hintId` is no longer needed. The component handles associating the hint text, label and input
+  automatically. If you wish to have a unique `id`, you can add one onto the `FormField` element.
+- `hintText` is no longer a valid prop. Use `FormField.Hint` sub component.
```
Comment (josh-bagwell) [inline modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:330]: is "sub component" hyphenated? I'm not sure. 

```suggestion
- `hintText` is no longer a valid prop. Use `FormField.Hint` sub-component.
```

File: modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx [hunk 11]
```diff
@@ -226,50 +228,121 @@ The following changes have been made to the API:
...
-`orientation="horizontalStart"` if you're using the string literal values.
+`orientation="horizontalStart"` if you're using the string literal values. It will also handle
+updating your imports from `@workday/canvas-kit-preview-react/form-field` to
+`@workday/canvas-kit-react/form-field`.
 
 #### Form Field Group
 
-We've added a new component under `FromField` to allow users to group inputs without worrying about
-setting the `as` prop on the `FormField` component.
+We've added a new component to the `form-field` package to allow users to group inputs without
```
Comment (josh-bagwell) [inline modules/docs/mdx/12.0-UPGRADE-GUIDE.mdx:344]: Did you want this hyphenated or like the above? 
```suggestion
We've added a new component to the `FormField` package to allow users to group inputs without
```

File: modules/labs-react/search-form/lib/SearchForm.tsx [hunk 12]
```diff
@@ -464,8 +461,10 @@ export class SearchForm extends React.Component<SearchFormProps, SearchFormState
               clearButtonAriaLabel={clearButtonAriaLabel}
               labelId={labelId}
             >
-              <SearchInput
-                ref={this.inputRef}
+              <FormField.Input
+                as={SearchInput}
+                ref={this.inputRef as any}
+                cs={{maxWidth: grow ? '100%' : maxWidth}}
```
Comment (josh-bagwell) [inline modules/labs-react/search-form/lib/SearchForm.tsx:467]: Should we do this as a modifier or leave using the cs prop?

Comment (josh-bagwell) [reply modules/labs-react/search-form/lib/SearchForm.tsx:467]: Or would this be an API change

File: modules/react/form-field/stories/FormField.mdx [hunk 13]
```diff
@@ -1,137 +1,167 @@
...
-Form Field allows users to wrap input components to make them accessible and adds complementary
-labels and error messages.
+FormField allows users to wrap input components to make them accessible. You can customize the field
+by passing in `TextInput`, `Select`, `RadioGroup` and other form elements to `FormField.Input`.
 
 ## Installation
 
 ```sh
-yarn add @workday/canvas-kit-react
+yarn add @workday/canvas-kit-preview-react
```
Comment (josh-bagwell) [inline modules/react/form-field/stories/FormField.mdx:30]: Should this be `preview` or from `main`?

```suggestion
yarn add @workday/canvas-kit-react
```

File: modules/labs-react/search-form/lib/SearchForm.tsx (hunk 12 above)
Comment (mannycarrera4) [reply modules/labs-react/search-form/lib/SearchForm.tsx:467]: leaving this as is since it's a labs component and hasn't been converted to a stencil

Comment (josh-bagwell) [reply modules/labs-react/search-form/lib/SearchForm.tsx:467]: This is part of labs, so we are not updating this

**Review by josh-bagwell — APPROVED** [review]

**Pull request merged**